sdk/ml/azure-ai-ml/azure/ai/ml/_utils/_cache_utils.py

```diff
@@ -52,76 +91,183 @@ def _get_component_hash(cls, component: Component) -> str:
+
+        if is_on_disk_cache_enabled():
+            cache_contents_to_resolve = self._resolve_cache_contents_from_disk(cache_contents_to_resolve)
+
+        self._resolve_cache_contents(cache_contents_to_resolve, resolver=self._resolver)
+
+        self._fill_back_component_to_nodes(dict_of_nodes_to_resolve)
+
+    def register_node_to_resolve(self, node: BaseNode):
+        """Register a node with its component to resolve.
+        """
+        component = node._component  # pylint: disable=protected-access
+
+        # directly resolve node if the resolution involves no remote call
+        # so that no node will be registered when create_or_update a subgraph
+        if isinstance(component, str):
+            node._component = self._resolver(  # pylint: disable=protected-access
+                component,
+                azureml_type=AzureMLResourceType.COMPONENT
+            )
+            return
+        if component.id is not None:
+            node._component = component.id  # pylint: disable=protected-access
+            return
+
+        self._nodes_to_resolve.append(node)
 
     def resolve_nodes(self):
         """Resolve all dependent components with resolver and set resolved component arm id back to newly
         registered nodes. Registered nodes will be cleared after resolution.
         """
-        if self._hash_of_components_to_resolve:
-            self._resolve_components_without_in_memory_cache()
+        if not self._nodes_to_resolve:
+            return
 
-        for component_hash, nodes in self._nodes_to_apply.items():
-            for node in nodes:
-                resolution_result = self._component_resolution_cache[component_hash]
-
-                # hack: parallel.task.code won't be resolved for now, but parallel.task can be a reference
-                # to parallel._component.task, then it will change to an arm id after resolving parallel._component
-                # However, if we have avoided duplicate resolution with in-memory cache, such change won't happen,
-                # so we need to do it manually here
-                # The ideal solution should be done after PRS team decides how to handle parallel.task.code
-                from azure.ai.ml.entities import ParallelComponent, Parallel
-                resolved_component = self._component_cache[component_hash]
-                if isinstance(node, Parallel) and isinstance(resolved_component, ParallelComponent):
-                    origin_component: ParallelComponent = node._component  # pylint: disable=protected-access
-                    if id(node.task) == id(origin_component.task):
-                        node.task.code = resolved_component.task.code
-
-                node._component = resolution_result  # pylint: disable=protected-access
-        self._nodes_to_apply.clear()
+        # all nodes will be skipped on register_node_to_resolve when resolving subgraph
+        self._lock.acquire()
```

Comment:
Since cache reading is thread safe, maybe we can add this lock before and after cache writing part?

Reply:
Node resolution involves filling back and will change the state of nodes, e.g. hash of its inner component.

This is for concurrent pipeline job creation. Lock won't be acquired in subgraph resolution as all nodes will be skipped on calling `register_node_to_resolve`.
So, for example, if `node_a` is used in both `pipeline1` and `pipeline2` and we create the 2 pipeline jobs concurrently, `node_a._component.code` will be a local path before `_resolve_nodes` and an arm id after `_resolve_nodes`.